scalableminds/webknossos
# Support transforming the skeleton layer with affine and TSP transforms

This PR takes care of transforming skeleton nodes (in the shader and when interacting with the positions) when necessary. During saving, the original node positions are stored (i.e., without any applied transformations).
In contrast to data layers (color and segmentation), the skeleton layer doesn't have an own transforms property. Instead it is treated like a layer that has no transforms at all. If a dataset with transformed layers exist, these layers will be transformed according to their transformations while the skeleton layer has the identity transforms. When one of the layers that have transforms is toggled to be rendered natively, the inverse transform is applied to the skeleton layer. As a result, the nodes will be transformed the same way the other layers are transformed.

When using the front-end export of skeletons to NML with transforms, the currently rendered node positions are written while the active transformation is encoded in the parameters section like the following. The array notation is loosely based on the TrakEm format (that uses something like `<t2_patch oid=“33” width=“949.0” height=“901.0” transform=“matrix(1.0,0.0,-0.0,1.0,53.0,72.0)... />`).

Affine:
```xml
<parameters>
    <experiment name="C555_DIAMOND_2f" description="" organization="sample_organization" wkUrl="http://localhost:9000" />
    ...
    
    <transform type="affine" positionsAreTransformed="true" matrix="[2.3271632194519043,12.151433944702148,0.6425480246543884,-5994.43896484375,-12.078832626342773,2.238189697265625,1.314159631729126,5853.94580078125,-1.229456901550293,0.8777050375938416,-12.358572006225586,8174.52685546875,0,0,0,1]" />
    ...
</parameters>
```

TPS:
```xml
  <parameters>
    ...
    
    <transform type="thin_plate_spline" positionsAreTransformed="true">
      <correspondence source="[570.3021454931336,404.5548993008739,502.2248079151061]" target="[568.1197424615385,528.0132258461539,1622.1119725846156]" />
      ...
  </parameters>
....
```

If we decide in the future, that we always want to store the untransformed positions in the NML, we can do so and then change `positionsAreTransformed="true"` to false.

### URL of deployed dev instance (used for testing):
- https://transformskeletons.webknossos.xyz/

### Steps to test:
- create a new annotation for the C555_DIAMOND_2f-TPS dataset
- note that the skeleton layer in the left sidebar will have an icon representing whether it is rendered with or without transforms
- display the C555_DIAMOND_2f layer natively --> the skeleton layer will now be rendered with transforms
- use the download trees feature in the trees tab
  - there is one normal download (without transforms) and
  - a download with transforms
- the skeleton tool should be disabled now
- test the merger mode
- also test skeleton annotation in an nd dataset (without transforms)

### TODOs:
- [x] Disable skeleton annotation when the layer is transformed. otherwise, slightly weird behavior can occur, because the clicked position can slightly differ from the stored position
  - this is because the node positions are stored as integers. this leads to inaccuracies when going from the clicked position to an untransformed position (that is truncated) and then back again to the transformed coordinate space.
  - ~[ ] alternatively, we could change the storage format of node positions to float vectors <-- let's do this only when necessary
- [x] when exporting a NML, do we want to encode which transformation system was applied during the import?
  - [x] which format? how to encode TPS?
  - [X] is it fair to not transform the coordinates when using the back-end download? yes, by default, you get the export as it's stored in the backend. if you need the transformed nodes, you can use the tree download which asks whether you want this.
- [X] how should we treat the back-end export? we could leave it as is and argue that omitted transformation metadata in the NML encode that the node positions were exported without any transforms. then, it would be semantically correct.
- ~[ ] should the node property "position" in the DB be renamed to "transformedPosition" (as the front-end did it?)~

### Issues:
- contributes to #7270

------
(Please delete unneeded items, merge only when none are left open)
